# Question 1




## (a)

The training error will increase as k varies from 1 to n, however, this increase isn't linear. The training error when k = 1 is going to be zero because each time the nearest neighbour of a node will be the point itself since its distance from itself will be zero, and the error will be the highest when k = n because the answer will remain the same for all data points.



## (b)

The generaisation error will decrease and then increase from k = 1 to n. Due to the overlapping nature of the data points, when k = 1 or a smaller number, there's a high possibility of the final class getting influenced by nearby noise points. As k gets larger, the error will reduce because the k nearest neighbours will start to return more accurate results. However, as k approaches n, the error increases again because the output can be either class with a 50/50 probability. The graph can be found [here](https://drive.google.com/file/d/1527x8CRLY_hj3TMJWodpNKPR-awPWSlq/view?usp=sharing).

<!-- <p><a href="https://drive.google.com/file/d/1527x8CRLY_hj3TMJWodpNKPR-awPWSlq/view?usp=sharing"><img src="https://drive.google.com/file/d/1527x8CRLY_hj3TMJWodpNKPR-awPWSlq/view?usp=sharing"></a></p> -->




## (c)

k-NN is an undesireable algorithm when the dimensionality of the input is high because:
* As mentioned in class, all the points lie on the surface of a unit hypersphere, which implies that the Eucledian distance between the query point and the other data points start to become very close in value, i.e., the closest and the farthest points have the same distance, and the term "nearest neighbours" starts to lose sense. Hence, choosing the k nearest neighbours based on distance makes no sense in higher dimensions and we start to get erroneous outputs
* Calculating distances in a Euclidean space makes sense for smaller dimensions but as we move to higher dimensions, it would make more sense to do the calculations in a probabilistic space
* In higher dimensions, the k-NN algorithm will also start to get more computationally expensive because calculating the Eucledean distance might become a large task which will increase computation time



## (d)

No, a univariate decision tree cannot function exactly like a 1-NN. To justify this, we must take into conosideration the Voronoi diagram described in class. It divides the 2D plane into regions whose edges are the decision boundaries obtained from the training examples. It is evident from the diagram that the edges can be in any direction and don't necessarily have to be parallel to the 2D axes. However, with a decision tree of the given format, all the decision boundaries will be parallel to the axes. This is clearly very different from the Voronoi diagram and will incorrectly classify a lot of points, and hence it cannot be used to function like a 1-NN.

# Question 2

## (a)

In the given question, to fit a Gaussian to the two classes using maximum likelihood, we just need to find $\mu$ which is just the mean of the given data. The fact that $\mu$ will be the mean of the given data can be easily proved by performing the following steps:

1. Take the logarithm of the maximum likelihood formula
2. Differentiate the given formula with respect to $\mu$
3. Set the above equation to zero since we need to maximise the likelihood and solve for $\mu$
4. Solving for $\mu$ will lead us to the answer which turns out to be the mean of the given data

$\mu = \frac{\sum_{i=1}^{n} x_i}{n}$
Note: the sigma values have been directly copied from the assignment document.


Thus, for class 1:\
$$
\mu = 0.26 \\
\sigma^2 = 0.0149
$$\
And for class 2:\
$$
\mu = 0.8625 \\
\sigma^2 = 0.0092
$$

Class probability is the probability of a point belonging to either class.
$$
p(c_i) = \frac{N_i}{\sum_{j=1}^{n} N_j}
$$

Where, $c_i$ is the class $i$ and $N_i$ is the size of $c_i$.

Thus, for class 1:\
$$
p(c_1) = \frac{10}{10+4}\\
p(c_1) = 0.714
$$\
And for class 2:\
$$
p(c_2) = \frac{4}{10+4}\\
p(c_2) = 0.286
$$

The Bayes' theorem:
$$
p(c_i|x) = \frac{p(x|c_i)p(c_i)}{p(x)}\\
p(c_i|x) = \frac{p(x|c_i)p(c_i)}{\sum_{j=1}^{n} p(x|c_j)p(c_j)}
$$

And the Gaussian formula:
$$
p(x|c_i) = \frac{1}{\sqrt{2\pi \sigma_i^{2}}} \exp \left(-\frac{(x -\mu_i)^2}{2\sigma_i^2} \right)
$$

Thus for $p(c_1|0.6)$:
$$
p(c_1|0.6) = \frac{p(0.6|c_1)p(c_1)}{p(0.6|c_1)p(c_1)+p(0.6|c_2)p(c_2)}\\
\\
p(0.6|c_1) = \frac{1}{\sqrt{2\pi * 0.0149}} \exp \left(-\frac{(0.6 - 0.26)^2}{2*0.0149} \right) = 0.0675\\
p(0.6|c_2) = \frac{1}{\sqrt{2\pi * 0.0092}} \exp \left(-\frac{(0.6 - 0.8625)^2}{2*0.0092} \right) = 0.0983\\
p(c_1|0.6) = 0.6316
$$

                                                               





## (b)

### Training Examples
| goal | football | golf | defence | offence | wicket | office | strategy | Sports or Not? |
|------|----------|------|---------|---------|--------|--------|----------|----------------|
|    1 |        0 |    1 |       1 |       1 |      0 |      1 |        1 |              0 |
|    0 |        0 |    0 |       1 |       0 |      0 |      1 |        1 |              0 |
|    1 |        0 |    0 |       1 |       1 |      0 |      1 |        0 |              0 |
|    0 |        1 |    0 |       0 |       1 |      1 |      0 |        1 |              0 |
|    0 |        0 |    0 |       1 |       1 |      0 |      1 |        1 |              0 |
|    0 |        0 |    0 |       1 |       1 |      0 |      0 |        1 |              0 |
|    1 |        1 |    0 |       0 |       0 |      0 |      0 |        0 |              1 |
|    0 |        0 |    1 |       0 |       0 |      0 |      0 |        0 |              1 |
|    1 |        1 |    0 |       1 |       0 |      0 |      0 |        0 |              1 |
|    1 |        1 |    0 |       1 |       0 |      0 |      0 |        1 |              1 |
|    1 |        1 |    0 |       1 |       1 |      0 |      0 |        0 |              1 |
|    0 |        0 |    0 |       1 |       0 |      1 |      0 |        0 |              1 |

### Look Up Tables

| politics                        	| sports                          	|
|---------------------------------	|---------------------------------	|
| <img width=250/>                  | <img width=250/>                  |
| $p(goal=1|sports=0) = 1/3$     	| $p(goal=1|sports=1) = 2/3$     	|
| $p(goal=0|sports=0) = 2/3$     	| $p(goal=0|sports=1) = 1/3$     	|
| $p(football=1|sports=0) = 1/6$ 	| $p(football=1|sports=1) = 2/3$ 	|
| $p(football=0|sports=0) = 5/6$ 	| $p(football=0|sports=1) = 1/3$ 	|
| $p(golf=1|sports=0) = 1/6$     	| $p(golf=1|sports=1) = 1/6$     	|
| $p(golf=0|sports=0) = 5/6$     	| $p(golf=0|sports=1) = 5/6$     	|
| $p(defence=1|sports=0) = 5/6$  	| $p(defence=1|sports=1) = 2/3$  	|
| $p(defence=0|sports=0) = 1/6$  	| $p(defence=0|sports=1) = 1/3$  	|
| $p(offence=1|sports=0) = 5/6$  	| $p(offence=1|sports=1) = 1/6$  	|
| $p(offence=0|sports=0) = 1/6$  	| $p(offence=0|sports=1) = 5/6$  	|
| $p(wicket=1|sports=0) = 1/6$   	| $p(wicket=1|sports=1) = 1/6$   	|
| $p(wicket=0|sports=0) = 5/6$   	| $p(wicket=0|sports=1) = 5/6$   	|
| $p(office=1|sports=0) = 2/3$   	| $p(office=1|sports=1) = 0/1$   	|
| $p(office=0|sports=0) = 1/3$   	| $p(office=0|sports=1) = 1/1$   	|
| $p(strategy=1|sports=0) = 5/6$ 	| $p(strategy=1|sports=1) = 1/6$ 	|
| $p(strategy=0|sports=0) = 1/6$ 	| $p(strategy=0|sports=1) = 5/6$ 	|



$$
p(politics) = \frac{6}{12} = 0.5\\
p(sports) = \frac{6}{12} = 0.5
$$


Given,\
$x = (1, 0, 0, 1, 1, 1, 1, 0)$

$
p(politics|x) = \frac{p(x|sports=0)p(politics)}{p(x|sports=0)p(politics) + p(x|sports=1)p(sports)}\\
$\
Now,\
$
p(x|sports=0)p(politics) = p(goal=1|sports=0) \times p(football=0|sports=0) \times p(golf=0|sports=0) \times p(defence=1|sports=0) \times p(offence=1|sports=0) \times p(wicket=1|sports=0) \times p(office=1|sports=0) \times p(strategy=0|sports=0) \times p(politics)
$\
$p(x|sports=0) = 0.00149$


To find the probability, we need to find $p(x|sports=1)$.

Now,\
$
p(x|sports=1)p(sports) = p(goal=1|sports=1) \times p(football=0|sports=1) \times p(golf=0|sports=1) \times p(defence=1|sports=1) \times p(offence=1|sports=1) \times p(wicket=1|sports=1) \times p(office=1|sports=1) \times p(strategy=0|sports=1) \times p(sports)
$\

$p(x|sports=1) = 0.0$

Thus,
$
p(politics|x) = \frac{0.00149}{0.00149 + 0.0} = 1
$


# Question 3

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


## (a)

In [3]:
# FoML Assign 1 Code Skeleton
# Please use this outline to implement your decision tree. You can add any code around this.

import csv
import numpy as np

# Enter You Name Here
myname = "Soumi-Chakraborty"

class Node():
  def __init__(self, left=None, right=None, attribute_index=None, threshold=None, label=None, isLeaf=False):
    self.left = left
    self.right = right
    self.attribute_index = attribute_index
    self.threshold = threshold
    self.label = label
    self.isLeaf = isLeaf

# Implement your decision tree below
class DecisionTree():

    def getInfoGain(self, node, left, right):
      infoGain = self.entropy(node) - (np.shape(left)[0] / np.shape(node)[0]) * self.entropy(left) - (np.shape(right)[0] / np.shape(node)[0]) * self.entropy(right)  
      return infoGain

    def entropy(self, y):
      #  since type can only be 0 or 1
      p0 = np.shape(y[y == 0])[0]/np.shape(y)[0]
      p1 = np.shape(y[y == 1])[0]/np.shape(y)[0]
      if p0 == 0 or p1 == 0:
        return 0
      entropy = -p0 * np.log2(p0) - p1 * np.log2(p1)
      return entropy
      

    def learn(self, training_set):
        # implement this function
        X = training_set[:, :-1] # all rows, all columns except the last
        y = training_set[:, -1]  # all rows, only the last column
        y = y.reshape(y.shape[0], 1)

        # purity check
        check = y[y == 1].shape[0]
        
        # split if max depth not reached
        if check != 0 and check != y.shape[0]:
          # get best split
          splitDetails = {
              "infoGain" : None, 
              "leftTrainingSet": None, 
              "rightTrainingSet": None, 
              "attributeIndex" : None, 
              "threshold": None
          }
          
          # check over all the attributes
          bestInfoGain = -float("inf")
          for feature in range(X.shape[1]):
            values = training_set[:, feature] 
            thersholds = np.unique(values)
            # compare all possible splits and choose the best one
            for thershold in thersholds:
              left_training_set = training_set[np.where(training_set[:, feature] <= thershold)]
              right_training_set = training_set[np.where(training_set[:, feature] > thershold)]
              if np.shape(left_training_set)[0] > 0 and np.shape(right_training_set)[0] > 0:
                yLeft = left_training_set[:, -1].reshape(np.shape(left_training_set)[0], 1)
                yRight = right_training_set[:, -1].reshape(np.shape(right_training_set)[0], 1)

                # get info gain
                infoGain = self.getInfoGain(y, yLeft, yRight)
                # update split
                if infoGain > bestInfoGain:
                  bestInfoGain = infoGain
                  splitDetails = {
                    "infoGain" : infoGain, 
                    "leftTrainingSet": left_training_set, 
                    "rightTrainingSet": right_training_set, 
                    "attributeIndex" : feature, 
                    "threshold": thershold
                    }

          # check if node is impure and recurse
          if splitDetails["infoGain"] != None and splitDetails["infoGain"] > 0:
            # repeat for left side
            left = self.learn(splitDetails["leftTrainingSet"])
            # repeat for left side
            right = self.learn(splitDetails["rightTrainingSet"])
            # return new node
            return Node(left, right, splitDetails["attributeIndex"], splitDetails["threshold"])
        
        # if the control reaches here, the node under consideration is a leaf node
        zeros = np.shape(y[y == 0])[0]
        ones = np.shape(y[y == 1])[0]
        if zeros > ones:
          return Node(label=0, isLeaf=True)
        else:
          return Node(label=1, isLeaf=True)

    def classify(self, root, test_instance):
        if root.isLeaf == True:
          return root.label
        if test_instance[root.attribute_index] <= root.threshold:
          return self.classify(root.left, test_instance)
        else:
          return self.classify(root.right, test_instance)


def run_decision_tree():

    # Load data set
    # data_set_path = "/content/drive/MyDrive/Acad_Stuff/Course Work/Semester 5/FoML/Ass1/resources/wine-dataset.csv"
    data_set_path = "wine-dataset.csv"
    # with open("wine-dataset.csv") as f:
    with open(data_set_path) as f:
        next(f, None)
        data = [tuple(line) for line in csv.reader(f, delimiter=",")]
    print("Number of records: %d" % len(data))
    
    # Split training/test sets
    # You need to modify the following code for cross validation.
    K = 10
    training_set = [x for i, x in enumerate(data) if i % K != 9]
    test_set = [x for i, x in enumerate(data) if i % K == 9]

    training_set = np.asarray(training_set, dtype=np.float32)
    test_set = np.asarray(test_set, dtype=np.float32)

    # print(np.shape(training_set))
    
    tree = DecisionTree()
    root = Node()
    # Construct a tree using training set
    root = tree.learn(training_set)

    # Classify the test set using the tree we just constructed
    results = []
    for instance in test_set:
        result = tree.classify(root, instance[:-1])
        # print(result)
        results.append( result == int(instance[-1]))

    
    # Accuracy
    accuracy = float(results.count(True))/float(len(results))
    print("accuracy: %.4f" % accuracy)       
    

    # Writing results to a file (DO NOT CHANGE)
    f = open(myname+"result.txt", "w")
    f.write("accuracy: %.4f" % accuracy)
    f.close()


if __name__ == "__main__":
    run_decision_tree()

Number of records: 4898
accuracy: 0.8221


## (b)

In [5]:
# FoML Assign 1 Code Skeleton
# Please use this outline to implement your decision tree. You can add any code around this.

import csv
import numpy as np

# Enter You Name Here
myname = "Soumi-Chakraborty"

class Node():
  def __init__(self, left=None, right=None, attribute_index=None, threshold=None, label=None, isLeaf=False):
    self.left = left
    self.right = right
    self.attribute_index = attribute_index
    self.threshold = threshold
    self.label = label
    self.isLeaf = isLeaf

# Implement your decision tree below
class DecisionTree():
    

    def getInfoGain(self, node, left, right):
      infoGain = self.entropy(node) - (np.shape(left)[0] / np.shape(node)[0]) * self.entropy(left) - (np.shape(right)[0] / np.shape(node)[0]) * self.entropy(right)  
      return infoGain

    def entropy(self, y):
      #  since type can only be 0 or 1
      p0 = np.shape(y[y == 0])[0]/np.shape(y)[0]
      p1 = np.shape(y[y == 1])[0]/np.shape(y)[0]
      if p0 == 0 or p1 == 0:
        return 0
      entropy = -p0 * np.log2(p0) - p1 * np.log2(p1)
      return entropy
      

    def learn(self, training_set):
        X = training_set[:, :-1] # all rows, all columns except the last
        y = training_set[:, -1]  # all rows, only the last column
        y = y.reshape(y.shape[0], 1)

        # purity check
        check = y[y == 1].shape[0]
        
        if check != 0 and check != y.shape[0]:
          # get best split
          splitDetails = {
              "infoGain" : None, 
              "leftTrainingSet": None, 
              "rightTrainingSet": None, 
              "attributeIndex" : None, 
              "threshold": None
          }
          
          # check over all the attributes
          bestInfoGain = -float("inf")
          for feature in range(X.shape[1]):
            values = training_set[:, feature] 
            thersholds = np.unique(values)
            # compare all possible splits and choose the best one
            for thershold in thersholds:
              # divide the training set based on the threshold
              left_training_set = training_set[np.where(training_set[:, feature] <= thershold)]
              right_training_set = training_set[np.where(training_set[:, feature] > thershold)]
              if np.shape(left_training_set)[0] > 0 and np.shape(right_training_set)[0] > 0:
                yLeft = left_training_set[:, -1].reshape(np.shape(left_training_set)[0], 1)
                yRight = right_training_set[:, -1].reshape(np.shape(right_training_set)[0], 1)

                # get info gain
                infoGain = self.getInfoGain(y, yLeft, yRight)
                # update split
                if infoGain > bestInfoGain:
                  bestInfoGain = infoGain
                  splitDetails = {
                    "infoGain" : infoGain, 
                    "leftTrainingSet": left_training_set, 
                    "rightTrainingSet": right_training_set, 
                    "attributeIndex" : feature, 
                    "threshold": thershold
                    }

          # check if node is impure 
          if splitDetails["infoGain"] != None and splitDetails["infoGain"] > 0:
            # repeat for left side
            left = self.learn(splitDetails["leftTrainingSet"])
            # repeat for left side
            right = self.learn(splitDetails["rightTrainingSet"])
            # return new node
            return Node(left, right, splitDetails["attributeIndex"], splitDetails["threshold"])
        
        # if the control reaches here, the node under consideration is a leaf node
        zeros = np.shape(y[y == 0])[0]
        ones = np.shape(y[y == 1])[0]
        if zeros > ones:
          return Node(label=0, isLeaf=True)
        else:
          return Node(label=1, isLeaf=True)

    def classify(self, root, test_instance):
        if root.isLeaf == True:
          return root.label
        if test_instance[root.attribute_index] <= root.threshold:
          return self.classify(root.left, test_instance)
        else:
          return self.classify(root.right, test_instance)

def run_decision_tree():

    # Load data set
    # data_set_path = "/content/drive/MyDrive/Acad_Stuff/Course Work/Semester 5/FoML/Ass1/resources/wine-dataset.csv"
    data_set_path = "wine-dataset.csv"

    # with open("wine-dataset.csv") as f:
    with open(data_set_path) as f:
        next(f, None)
        data = [tuple(line) for line in csv.reader(f, delimiter=",")]
    print("Number of records: %d" % len(data))
    
    # Split training/test sets
    # You need to modify the following code for cross validation.
    K = 10
    finalAccuracy = 0
    for j in range(K):
      training_set = [x for i, x in enumerate(data) if i % K != j]
      test_set = [x for i, x in enumerate(data) if i % K == j]

      training_set = np.asarray(training_set, dtype=np.float32)
      test_set = np.asarray(test_set, dtype=np.float32)

      # print(np.shape(training_set), np.shape(test_set))
      
      tree = DecisionTree()
      root = Node()
      # Construct a tree using training set
      root = tree.learn(training_set)

      # Classify the test set using the tree we just constructed
      results = []
      for instance in test_set:
          result = tree.classify(root, instance[:-1])
          # print(result)
          results.append( result == int(instance[-1]))

      
      # Accuracy
      accuracy = float(results.count(True))/float(len(results))
      print("accuracy: %.4f" % accuracy)
      finalAccuracy = finalAccuracy + accuracy


    finalAccuracy = finalAccuracy/K
    accuracy = finalAccuracy
    print("final accuracy: %.4f" % accuracy)
    

    # Writing results to a file (DO NOT CHANGE)
    f = open(myname+"result.txt", "w")
    f.write("accuracy: %.4f" % finalAccuracy)
    # f.write("final accuracy: %.4f" % finalAccuracy)
    f.close()


if __name__ == "__main__":
    run_decision_tree()

Number of records: 4898
accuracy: 0.8388
accuracy: 0.8633
accuracy: 0.8367
accuracy: 0.8408
accuracy: 0.8551
accuracy: 0.8041
accuracy: 0.8347
accuracy: 0.8429
accuracy: 0.8425
accuracy: 0.8221
final accuracy: 0.8381


## (c)

### Improvement method 1: Gini Index

In [6]:
# FoML Assign 1 Code Skeleton
# Please use this outline to implement your decision tree. You can add any code around this.

import csv
import numpy as np

# Enter You Name Here
myname = "Soumi-Chakraborty"

class Node():
  def __init__(self, left=None, right=None, attribute_index=None, threshold=None, label=None, isLeaf=False):
    self.left = left
    self.right = right
    self.attribute_index = attribute_index
    self.threshold = threshold
    self.label = label
    self.isLeaf = isLeaf

# Implement your decision tree below
class DecisionTree():

    def getInfoGain(self, node, left, right):
      # infoGain = self.entropy(node) - (len(left) / len(node)) * self.entropy(left) - (len(right) / len(node)) * self.entropy(right) 
      infoGain = self.giniIndex(node) - ((np.shape(left)[0] / np.shape(node)[0]) * self.giniIndex(left) + (np.shape(right)[0] / np.shape(node)[0]) * self.giniIndex(right))

      return infoGain

    def giniIndex(self, y):
      #  since type can only be 0 or 1
      p0 = np.shape(y[y == 0])[0]/np.shape(y)[0]
      p1 = np.shape(y[y == 1])[0]/np.shape(y)[0]
      gini = 1 - p0**2 - p1**2
      return gini
      

    def learn(self, training_set, depth):
        # implement this function
        X = training_set[:, :-1] # all rows, all columns except the last
        y = training_set[:, -1]  # all rows, only the last column
        y = y.reshape(y.shape[0], 1)

        # purity check
        check = y[y == 1].shape[0]
        
        # # split if max depth not reached
        if check != 0 and check != y.shape[0]:
          # get best split
          splitDetails = {
              "infoGain" : None, 
              "leftTrainingSet": None, 
              "rightTrainingSet": None, 
              "attributeIndex" : None, 
              "threshold": None
          }
          
          # check over all the attributes
          bestInfoGain = -float("inf")
          for feature in range(X.shape[1]):
            values = training_set[:, feature] 
            thersholds = np.unique(values)
            # compare all possible splits and choose the best one
            for thershold in thersholds:
              # divide the training set based on the threshold
              left_training_set = training_set[np.where(training_set[:, feature] <= thershold)]
              right_training_set = training_set[np.where(training_set[:, feature] > thershold)]
              if np.shape(left_training_set)[0] > 0 and np.shape(right_training_set)[0] > 0:
                yLeft = left_training_set[:, -1].reshape(np.shape(left_training_set)[0], 1)
                yRight = right_training_set[:, -1].reshape(np.shape(right_training_set)[0], 1)

                # get info gain
                infoGain = self.getInfoGain(y, yLeft, yRight)
                # update split
                if infoGain > bestInfoGain:
                  bestInfoGain = infoGain
                  splitDetails = {
                    "infoGain" : infoGain, 
                    "leftTrainingSet": left_training_set, 
                    "rightTrainingSet": right_training_set, 
                    "attributeIndex" : feature, 
                    "threshold": thershold
                    }

          # check if node is impure 
          if splitDetails["infoGain"] != None and splitDetails["infoGain"] > 0:
            # repeat for left side
            left = self.learn(splitDetails["leftTrainingSet"], depth + 1)
            # repeat for left side
            right = self.learn(splitDetails["rightTrainingSet"], depth + 1)
            # return new node
            return Node(left, right, splitDetails["attributeIndex"], splitDetails["threshold"])
        
        # if the control reaches here, the node under consideration is a leaf node
        zeros = np.shape(y[y == 0])[0]
        ones = np.shape(y[y == 1])[0]
        if zeros > ones:
          return Node(label=0, isLeaf=True)
        else:
          return Node(label=1, isLeaf=True)

    def classify(self, root, test_instance):
        result = 0 # baseline: always classifies as 0
        if root.isLeaf == True:
          return root.label
        if test_instance[root.attribute_index] <= root.threshold:
          return self.classify(root.left, test_instance)
        else:
          return self.classify(root.right, test_instance)

        # return result

def run_decision_tree():

    # Load data set
    # data_set_path = "/content/drive/MyDrive/Acad_Stuff/Course Work/Semester 5/FoML/Ass1/resources/wine-dataset.csv"
    data_set_path = "wine-dataset.csv"

    # with open("wine-dataset.csv") as f:
    with open(data_set_path) as f:
        next(f, None)
        data = [tuple(line) for line in csv.reader(f, delimiter=",")]
    print("Number of records: %d" % len(data))
    
    # Split training/test sets
    # You need to modify the following code for cross validation.
    K = 10
    finalAccuracy = 0
    for j in range(K):
      training_set = [x for i, x in enumerate(data) if i % K != j]
      test_set = [x for i, x in enumerate(data) if i % K == j]

      training_set = np.asarray(training_set, dtype=np.float32)
      test_set = np.asarray(test_set, dtype=np.float32)


      # print(np.shape(training_set), np.shape(test_set))
      
      tree = DecisionTree()
      root = Node()
      # Construct a tree using training set
      root = tree.learn(training_set, 0)

      # Classify the test set using the tree we just constructed
      results = []
      for instance in test_set:
          result = tree.classify(root, instance[:-1])
          # print(result)
          results.append( result == int(instance[-1]))

      
      # Accuracy
      accuracy = float(results.count(True))/float(len(results))
      print("accuracy: %.4f" % accuracy)
      finalAccuracy = finalAccuracy + accuracy


    finalAccuracy = finalAccuracy/K
    accuracy = finalAccuracy
    print("final accuracy: %.4f" % accuracy)
    

    # Writing results to a file (DO NOT CHANGE)
    f = open(myname+"result.txt", "w")
    f.write("accuracy: %.4f" % finalAccuracy)
    # f.write("final accuracy: %.4f" % finalAccuracy)
    f.close()


if __name__ == "__main__":
    run_decision_tree()

Number of records: 4898
accuracy: 0.8571
accuracy: 0.8633
accuracy: 0.8286
accuracy: 0.8469
accuracy: 0.8571
accuracy: 0.8306
accuracy: 0.8286
accuracy: 0.8612
accuracy: 0.8446
accuracy: 0.8303
final accuracy: 0.8448


### Improvement method 2: Pruning

In [7]:
# FoML Assign 1 Code Skeleton
# Please use this outline to implement your decision tree. You can add any code around this.

import csv
import numpy as np

# Enter You Name Here
myname = "Soumi-Chakraborty"
MAX_DEPTH = 12

class Node():
  def __init__(self, left=None, right=None, attribute_index=None, threshold=None, infoGain=None, label=None, depth=None, isLeaf=False):
    self.left = left
    self.right = right
    self.attribute_index = attribute_index
    self.threshold = threshold
    self.infoGain = infoGain
    self.label = label
    self.depth = depth
    self.isLeaf = isLeaf

# Implement your decision tree below
class DecisionTree():

    def getInfoGain(self, node, left, right):
      infoGain = self.giniIndex(node) - ((np.shape(left)[0] / np.shape(node)[0]) * self.giniIndex(left) + (np.shape(right)[0] / np.shape(node)[0]) * self.giniIndex(right))
      return infoGain

    def giniIndex(self, y):
      #  since type can only be 0 or 1
      p0 = np.shape(y[y == 0])[0]/np.shape(y)[0]
      p1 = np.shape(y[y == 1])[0]/np.shape(y)[0]
      gini = 1 - p0**2 - p1**2
      return gini
      

    def learn(self, training_set, depth):
        # implement this function
        X = training_set[:, :-1] # all rows, all columns except the last
        y = training_set[:, -1]  # all rows, only the last column
        y = y.reshape(y.shape[0], 1)

        # purity check
        check = y[y == 1].shape[0]
        
        # # split if max depth not reached
        if check != 0 and check != y.shape[0]:
          # get best split
          splitDetails = {
              "infoGain" : None, 
              "leftTrainingSet": None, 
              "rightTrainingSet": None, 
              "attributeIndex" : None, 
              "threshold": None
          }
          
          # check over all the attributes
          bestInfoGain = -float("inf")
          for feature in range(X.shape[1]):
            values = training_set[:, feature] 
            thersholds = np.unique(values)
            # compare all possible splits and choose the best one
            for thershold in thersholds:
              # divide the training set based on the threshold
              left_training_set = training_set[np.where(training_set[:, feature] <= thershold)]
              right_training_set = training_set[np.where(training_set[:, feature] > thershold)]
              if np.shape(left_training_set)[0] > 0 and np.shape(right_training_set)[0] > 0:
                yLeft = left_training_set[:, -1].reshape(np.shape(left_training_set)[0], 1)
                yRight = right_training_set[:, -1].reshape(np.shape(right_training_set)[0], 1)

                # get info gain
                infoGain = self.getInfoGain(y, yLeft, yRight)
                # print(infoGain)
                # update split
                if infoGain > bestInfoGain:
                  bestInfoGain = infoGain
                  splitDetails = {
                    "infoGain" : infoGain, 
                    "leftTrainingSet": left_training_set, 
                    "rightTrainingSet": right_training_set, 
                    "attributeIndex" : feature, 
                    "threshold": thershold
                    }

          # check if node is impure 

          if splitDetails["infoGain"] != None and splitDetails["infoGain"] > 0:
            # repeat for left side
            left = self.learn(splitDetails["leftTrainingSet"], depth + 1)
            # repeat for left side
            right = self.learn(splitDetails["rightTrainingSet"], depth + 1)
            
            # storing label of each node so we don't have to repeat the steps while pruning
            label = 0
            zeros = np.shape(y[y == 0])[0]
            ones = np.shape(y[y == 1])[0]
            if zeros > ones:
              label = 0
            else:
              label = 1
            
            # return new node
            return Node(left, right, splitDetails["attributeIndex"], splitDetails["threshold"], splitDetails["infoGain"], label, depth)
        
        # if the control reaches here, the node under consideration is a leaf node
        zeros = np.shape(y[y == 0])[0]
        ones = np.shape(y[y == 1])[0]
        if zeros > ones:
          # print("0")
          return Node(label=0, isLeaf=True)
        else:
          # print("1")
          return Node(label=1, isLeaf=True)

    def classify(self, root, test_instance):
        result = 0 # baseline: always classifies as 0
        if root.isLeaf == True:
          return root.label
        if test_instance[root.attribute_index] <= root.threshold:
          return self.classify(root.left, test_instance)
        else:
          return self.classify(root.right, test_instance)

        # return result

    # top-down post-pruning
    def prune(self, root):
      if (root.isLeaf == False and root.depth <= MAX_DEPTH):
        self.prune(root.left)
        self.prune(root.right)
      elif (root.isLeaf == False and root.infoGain * 100 >= 1):
        self.prune(root.left)
        self.prune(root.right)
      else:
        root.isLeaf = True
        root.left = None
        root.right = None
        root.infoGain = None


def run_decision_tree():

    # Load data set
    # data_set_path = "/content/drive/MyDrive/Acad_Stuff/Course Work/Semester 5/FoML/Ass1/resources/wine-dataset.csv"
    data_set_path = "wine-dataset.csv"

    # with open("wine-dataset.csv") as f:
    with open(data_set_path) as f:
        next(f, None)
        data = [tuple(line) for line in csv.reader(f, delimiter=",")]
    print("Number of records: %d" % len(data))
    
    # Split training/test sets
    # You need to modify the following code for cross validation.
    K = 10
    finalAccuracy = 0
    for j in range(K):
      training_set = [x for i, x in enumerate(data) if i % K != j]
      test_set = [x for i, x in enumerate(data) if i % K == j]

      training_set = np.asarray(training_set, dtype=np.float32)
      test_set = np.asarray(test_set, dtype=np.float32)


      # print(np.shape(training_set), np.shape(test_set))
      
      tree = DecisionTree()
      root = Node()
      # Construct a tree using training set
      root = tree.learn(training_set, 0)
      tree.prune(root)
      # print(root.infoGain)

      # Classify the test set using the tree we just constructed
      results = []
      for instance in test_set:
          result = tree.classify(root, instance[:-1])
          # print(result)
          results.append( result == int(instance[-1]))

      
      # Accuracy
      accuracy = float(results.count(True))/float(len(results))
      print("accuracy: %.4f" % accuracy)
      finalAccuracy = finalAccuracy + accuracy


    finalAccuracy = finalAccuracy/K
    accuracy = finalAccuracy
    print("final accuracy: %.4f" % accuracy)
    

    # Writing results to a file (DO NOT CHANGE)
    f = open(myname+"result.txt", "w")
    f.write("accuracy: %.4f" % finalAccuracy)
    # f.write("final accuracy: %.4f" % finalAccuracy)
    f.close()


if __name__ == "__main__":
    run_decision_tree()

Number of records: 4898
accuracy: 0.8571
accuracy: 0.8633
accuracy: 0.8286
accuracy: 0.8469
accuracy: 0.8571
accuracy: 0.8265
accuracy: 0.8286
accuracy: 0.8653
accuracy: 0.8466
accuracy: 0.8303
final accuracy: 0.8450


Accuracy without 10-fold cross validation: 82.21%\
Accuracy with 10-fold cross validation: 83.81%\
Accuracy with just Gini index instead of entropy: 84.48%\
Accuracy with Gini index and pruning: 84.50%

The improvement methods I implemented were to replace entropy with the gini index, and then do top-down pruning alongside gini. Calculating the gini index is less computationally expensive than calculating the entropy. The top-down pruning only starts to prune the tree after a certain depth (arbritarity chosen to be 12) and prunes if the information gain of any node is below a  particular thershold (arbitrarily chosen to be 1%) and these combined methods led to an increase in accuracy because it removes some overfitting subtrees.
